torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -355,16 +373,26 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
         } else if (message.isResponse()) {
           auto id = message.id();
           std::shared_ptr<FutureMessage> fm = nullptr;
+          std::chrono::milliseconds futureStartTime;
           {
             std::lock_guard<std::mutex> lock{futureMutex_};
-            fm = futures_[id];
+            std::tie(fm, futureStartTime) = futures_[id];
           }
           // Not holding lock on markCompleted as this could run callbacks that
           // call agent_->send
           fm->markCompleted(std::move(message));
           {
             std::lock_guard<std::mutex> lock{futureMutex_};
             futures_.erase(id);
+            // look up the corresponding future by its time out and request ID,
+            // and remove it from the timeouts map
+            auto& futuresAtTime = futureTimeouts_[futureStartTime];
+            futuresAtTime.erase(
```

Comment:
This code assumes that there is the desired future in the `futureTimeouts_` map, but this may not be the case when we implement removing futures that have timed out. Such a code path would get triggered when a future has timed out and been marked as so, but it later completes and we process it here. In cases like this, I think that we should just drop the future, since the user's request would have already been completed with a timed out exception.